File: dotnet-spring-equivalent/Controllers/MetricsController.cs
// <copyright file="MetricsController.cs" company="SpringJavaEquivalent">
// Copyright (c) 2024. All rights reserved.
// </copyright>

namespace SpringJavaEquivalent.Controllers;

using System.Diagnostics;
using System.Diagnostics.Metrics;
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("api/[controller]")]
public class MetricsController : ControllerBase
{
    private static readonly Meter Meter = new("SpringJavaEquivalent.Metrics", "1.0.0");
    private static readonly Counter<long> RequestCounter = Meter.CreateCounter<long>("custom_requests_total", "Total number of custom requests");
    private static readonly Histogram<double> ResponseTime = Meter.CreateHistogram<double>("custom_response_time", "Custom response time");
    private static readonly Random Random = new();

    [HttpGet("test")]
    public async Task<string> TestMetrics()
    {
        RequestCounter.Add(1);

        using var activity = new Activity("custom_operation");
        activity.Start();

        try
        {
            // Simulate some work
            await Task.Delay(Random.Next(50, 150));
            return "Metrics test completed";
        }
        finally
        {
            activity.Stop();
            ResponseTime.Record(activity.Duration.TotalMilliseconds);
        }
    }

    [HttpGet("counter")]
    public string IncrementCounter([FromQuery] int value = 1)
    {
        RequestCounter.Add(value);
        return $"Counter incremented by {value}";
    }

    [HttpGet("slow")]
    public async Task<string> SlowEndpoint()
    {
        RequestCounter.Add(1);

        using var activity = new Activity("slow_operation");
        activity.Start();

        try
        {
            // Simulate slow operation
            await Task.Delay(Random.Next(500, 1500));
            return "Slow operation completed";
        }
        finally
        {
            activity.Stop();
            ResponseTime.Record(activity.Duration.TotalMilliseconds);
        }
    }
}

File: src/main/java/hello/MetricsController.java
package hello;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Timer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.Random;


@RestController
@RequestMapping("/api/metrics")
public class MetricsController {

    private final Counter requestCounter;
    private final Timer responseTimer;
    private final Random random = new Random();

    @Autowired
    public MetricsController(MeterRegistry meterRegistry) {
        this.requestCounter = Counter.builder("custom_requests_total")
                .description("Total number of custom requests")
                .register(meterRegistry);
        
        this.responseTimer = Timer.builder("custom_response_time")
                .description("Custom response time")
                .register(meterRegistry);
    }

    @GetMapping("/test")
    public String testMetrics() {
        requestCounter.increment();
        
        Timer.Sample sample = Timer.start();
        try {
            // Simulate some work
            Thread.sleep(50L + random.nextInt(100));
            return "Metrics test completed";
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return "Metrics test interrupted";
        } finally {
            sample.stop(responseTimer);
        }
    }

    @GetMapping("/counter")
    public String incrementCounter(@RequestParam(defaultValue = "1") int value) {
        requestCounter.increment(value);
        return "Counter incremented by " + value;
    }

    @GetMapping("/slow")
    public String slowEndpoint() {
        requestCounter.increment();
        
        Timer.Sample sample = Timer.start();
        try {
            // Simulate slow operation
            Thread.sleep(500L + random.nextInt(1000));
            return "Slow operation completed";
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return "Slow operation interrupted";
        } finally {
            sample.stop(responseTimer);
        }
    }
}
